### ISSUE TITLE
Improve /docs/reference/hetzner-bare-metal-machine-template.md

### ISSUE THREAD

**Issue by pucilpet** [syself/cluster-api-provider-hetzner#209]
**Describe the solution you'd like**
/docs/reference/hetzner-bare-metal-machine-template.md and especially template.spec.hostSelector could probably use some references/examples.

Should we use GET/server object values from [hetzner webservice reference](https://robot.your-server.de/doc/webservice/en.html#server) for keys in the template.spec.hostSelector.matchExpressions.key ?

**Anything else you would like to add:**
I am also wondering whether there should be a cluster-template for HetznerBareMetalHost to make it clearer.

/kind proposal

**Comment by janiskemper** [comment]
Hi @pucilpet!
Quick answer to the hostSelcetor: You don't have to set it at all! You can use labels on your host and let a bm machine select a certain host by its labels. If not, then just any host is chosen. You see in the reference that the hostSelector is optional. You don't have to worry about it at all ;) 
Regarding the cluster template: Each host object has to be created in a highly specific way. We have described in the reference what fields have to be specified. We cannot create a cluster-template for this, as this wouldn't work out of the box. Everybody would have to adapt it. If you have another suggestion where we should have an example, then tell us! Here is a template, (which you have to adapt, of course):
```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: HetznerBareMetalHost
metadata:
  name: "bm-0"
spec:
  serverID: 1682566
  rootDeviceHints:
    wwn: "eui.0068475201b4egh2"
  maintenanceMode: false
  description: Test Machine 0
  ```

**Comment by pucilpet** [comment]
> Hi @pucilpet! Quick answer to the hostSelcetor: You don't have to set it at all! You can use labels on your host and let a bm machine select a certain host by its labels. If not, then just any host is chosen. You see in the reference that the hostSelector is optional. You don't have to worry about it at all ;) Regarding the cluster template: Each host object has to be created in a specific way. 

Thanks for your answer @janiskemper 

That is exactly what I am trying to avoin - having a server selected that is already being used for something else. :)

I understand your point regarding the cluster template. But it might be better to have some skeleton template than nothing at all. :)

Thanks again!

**Comment by pucilpet** [comment]
Also - from what I can see from /docs/reference/hetzner-bare-metal-machine-template.md - the template.spec.hostSelector is actually marked as required. :)

**Comment by janiskemper** [comment]
I think we can maybe put an example into the docs. 

I am happy that you understood the relationship between the objects now. Do you have any suggestion on how we can improve the documentation so that it is easier to understand?

And yes, you are absolutely right. hostSelector is marked as required, but it is not. As can be seen in the cluster templates, where we don't use it. 

Do you want to make a contribution to the docs? Then you could also update the hostSelector required field. We definitely welcome new contributors ;) If you have no suggestion on how to improve the docs, then I'll update the required=true field.

Anyway, thanks for finding the mistake in the docs!

**Comment by pucilpet** [comment]
Let me think about how to improve it and I will work on a PR.

**Comment by janiskemper** [comment]
Thank you so much @pucilpet! I'm looking forward to it!

**Comment by pucilpet** [comment]
One more thing that I noticed, probably might be a good idea to open a new issue?

In /docs/topics/preparation.md it is stated that 

> Preparing Hetzner Robot
> 
> Generate an SSH key. You can either upload it via Hetzner Robot UI, **or you can just rely on the controller to upload a key that it does not find in the robot API**. This is possible, as you have to store the public and private key together with the ssh key's name in a secret that the controller reads.

But when I added the ssh key only to a secret and not via Hetzner Robot UI, the hetznerbaremetalhosts stay in "preparing" state, never actually rebooting the servers.

Only after I added the SSH key via Hetzner Robot UI and named it "cluster" I was able to provision the baremetalhosts.

Is this something you have seen, too? Should I include in the docs PR that uploading the SSH key to Hetzner robot UI is required with the name cluster or maybe there needs to be some code change?

**Comment by janiskemper** [comment]
@pucilpet in theory it should work to upload ssh keys to robot API that are not yet registered, see https://github.com/syself/cluster-api-provider-hetzner/blob/9e78dcc13ee2075e257b27f8f10b98fd057e80b0/pkg/services/baremetal/host/host.go#L356.
Can you maybe have a look at the logs whether there are error messages related to this?

**Comment by janiskemper** [comment]
If this does not work there should be error logs. And yes, if you find some it would be good to open a new issue for this, as it has nothing to do anymore with the original topic of this issue here. Thank you!
